File: backend/utils/validation.ts
import type { Question } from '../../src/types';

// Quiz code format: 6 alphanumeric characters
const QUIZ_CODE_REGEX = /^[A-Z0-9]{6}$/;

// Team name: 1-50 characters, no leading/trailing whitespace
const TEAM_NAME_MIN_LENGTH = 1;
const TEAM_NAME_MAX_LENGTH = 50;

// Quiz validation
const QUIZ_TITLE_MIN_LENGTH = 1;
const QUIZ_TITLE_MAX_LENGTH = 200;
const MIN_QUESTIONS = 1;
const MAX_QUESTIONS = 100;

export function validateQuizCode(code: string): boolean {
  return QUIZ_CODE_REGEX.test(code);
}

export function generateQuizCode(): string {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let code = '';
  for (let i = 0; i < 6; i++) {
    code += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return code;
}

export function validateTeamName(name: string): {
  valid: boolean;
  error?: string;
} {
  if (!name || typeof name !== 'string') {
    return { valid: false, error: 'Team name is required' };
  }

  const trimmedName = name.trim();

  if (trimmedName.length < TEAM_NAME_MIN_LENGTH) {
    return { valid: false, error: 'Team name cannot be empty' };
  }

  if (trimmedName.length > TEAM_NAME_MAX_LENGTH) {
    return {
      valid: false,
      error: `Team name cannot exceed ${TEAM_NAME_MAX_LENGTH} characters`,
    };
  }

  return { valid: true };
}

export function validateQuizTitle(title: string): {
  valid: boolean;
  error?: string;
} {
  if (!title || typeof title !== 'string') {
    return { valid: false, error: 'Quiz title is required' };
  }

  const trimmedTitle = title.trim();

  if (trimmedTitle.length < QUIZ_TITLE_MIN_LENGTH) {
    return { valid: false, error: 'Quiz title cannot be empty' };
  }

  if (trimmedTitle.length > QUIZ_TITLE_MAX_LENGTH) {
    return {
      valid: false,
      error: `Quiz title cannot exceed ${QUIZ_TITLE_MAX_LENGTH} characters`,
    };
  }

  return { valid: true };
}

export function validateQuestions(questions: Question[]): {
  valid: boolean;
  error?: string;
} {
  if (!Array.isArray(questions)) {
    return { valid: false, error: 'Questions must be an array' };
  }

  if (questions.length < MIN_QUESTIONS) {
    return { valid: false, error: 'At least one question is required' };
  }

  if (questions.length > MAX_QUESTIONS) {
    return {
      valid: false,
      error: `Cannot exceed ${MAX_QUESTIONS} questions`,
    };
  }

  for (let i = 0; i < questions.length; i++) {
    const question = questions[i];

    if (!question.text || typeof question.text !== 'string') {
      return { valid: false, error: `Question ${i + 1}: Text is required` };
    }

    if (question.text.trim().length === 0) {
      return { valid: false, error: `Question ${i + 1}: Text cannot be empty` };
    }

    if (!question.correct || typeof question.correct !== 'string') {
      return { valid: false, error: `Question ${i + 1}: Correct answer is required` };
    }

    if (question.correct.trim().length === 0) {
      return { valid: false, error: `Question ${i + 1}: Correct answer cannot be empty` };
    }
  }

  return { valid: true };
}

export function validateAnswerSubmission(
  questionId: number,
  answer: string,
  totalQuestions: number,
): { valid: boolean; error?: string } {
  if (typeof questionId !== 'number' || questionId < 0) {
    return { valid: false, error: 'Invalid question ID' };
  }

  if (questionId >= totalQuestions) {
    return { valid: false, error: 'Question not found' };
  }

  if (typeof answer !== 'string' || answer.trim().length === 0) {
    return { valid: false, error: 'Answer cannot be empty' };
  }

  return { valid: true };
}

File: backend-ktor/src/main/kotlin/com/pubquiz/validation/Validation.kt
package com.pubquiz.validation

import com.pubquiz.models.Question
import kotlin.random.Random

// Constants
private val QUIZ_CODE_REGEX = Regex("^[A-Z0-9]{6}$")
private const val TEAM_NAME_MIN_LENGTH = 1
private const val TEAM_NAME_MAX_LENGTH = 50
private const val QUIZ_TITLE_MIN_LENGTH = 1
private const val QUIZ_TITLE_MAX_LENGTH = 200
private const val MIN_QUESTIONS = 1
private const val MAX_QUESTIONS = 100

data class ValidationResult(
    val valid: Boolean,
    val error: String? = null
)

object Validation {
    fun validateQuizCode(code: String): Boolean {
        return code.matches(QUIZ_CODE_REGEX)
    }
    
    fun generateQuizCode(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return (1..6)
            .map { chars[Random.nextInt(chars.length)] }
            .joinToString("")
    }
    
    fun validateTeamName(name: String?): ValidationResult {
        if (name == null || name.isEmpty()) {
            return ValidationResult(false, "Team name is required")
        }
        
        val trimmedName = name.trim()
        
        if (trimmedName.isEmpty()) {
            return ValidationResult(false, "Team name cannot be empty")
        }
        
        if (trimmedName.length > TEAM_NAME_MAX_LENGTH) {
            return ValidationResult(false, "Team name cannot exceed $TEAM_NAME_MAX_LENGTH characters")
        }
        
        return ValidationResult(true)
    }
    
    fun validateQuizTitle(title: String?): ValidationResult {
        if (title.isNullOrBlank()) {
            return ValidationResult(false, "Quiz title is required")
        }
        
        val trimmedTitle = title.trim()
        
        if (trimmedTitle.length < QUIZ_TITLE_MIN_LENGTH) {
            return ValidationResult(false, "Quiz title cannot be empty")
        }
        
        if (trimmedTitle.length > QUIZ_TITLE_MAX_LENGTH) {
            return ValidationResult(false, "Quiz title cannot exceed $QUIZ_TITLE_MAX_LENGTH characters")
        }
        
        return ValidationResult(true)
    }
    
    fun validateQuestions(questions: List<Question>?): ValidationResult {
        if (questions == null) {
            return ValidationResult(false, "Questions must be provided")
        }
        
        if (questions.size < MIN_QUESTIONS) {
            return ValidationResult(false, "At least one question is required")
        }
        
        if (questions.size > MAX_QUESTIONS) {
            return ValidationResult(false, "Cannot exceed $MAX_QUESTIONS questions")
        }
        
        questions.forEachIndexed { index, question ->
            if (question.text.isBlank()) {
                return ValidationResult(false, "Question ${index + 1}: Text cannot be empty")
            }
            
            if (question.correct.isBlank()) {
                return ValidationResult(false, "Question ${index + 1}: Correct answer cannot be empty")
            }
        }
        
        return ValidationResult(true)
    }
    
    fun validateAnswerSubmission(
        questionId: Int,
        answer: String?,
        totalQuestions: Int
    ): ValidationResult {
        if (questionId < 0) {
            return ValidationResult(false, "Invalid question ID")
        }
        
        if (questionId >= totalQuestions) {
            return ValidationResult(false, "Question not found")
        }
        
        if (answer.isNullOrBlank()) {
            return ValidationResult(false, "Answer cannot be empty")
        }
        
        return ValidationResult(true)
    }
}
